File: superstitioMain/java/superstitio/cards/lupa/SkillCard/cardManipulation/ZenState.java
package superstitio.cards.lupa.SkillCard.cardManipulation;

import basemod.cardmods.ExhaustMod;
import basemod.helpers.CardModifierManager;
import com.megacrit.cardcrawl.actions.common.ExhaustSpecificCardAction;
import com.megacrit.cardcrawl.cards.AbstractCard;
import com.megacrit.cardcrawl.characters.AbstractPlayer;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.monsters.AbstractMonster;
import superstitio.DataManager;
import superstitio.cards.lupa.LupaCard;
import superstitio.delayHpLose.RemoveDelayHpLoseBlock;
import superstitioapi.actions.AutoDoneInstantAction;
import superstitioapi.actions.ChoseCardFromHandCardSelectScreen;
import superstitioapi.utils.CardUtility;

public class ZenState extends LupaCard {
    public static final String ID = DataManager.MakeTextID(ZenState.class);

    public static final CardType CARD_TYPE = CardType.SKILL;

    public static final CardRarity CARD_RARITY = CardRarity.COMMON;

    public static final CardTarget CARD_TARGET = CardTarget.SELF;

    private static final int COST = 0;
    private static final int COST_UPGRADED_NEW = 0;
    private static final int BLOCK = 4;
    private static final int UPGRADE_BLOCK = 2;
    private static final int MAGIC = 1;

    public ZenState() {
        super(ID, CARD_TYPE, COST, CARD_RARITY, CARD_TARGET);
        this.setupMagicNumber(MAGIC);
        this.setupBlock(BLOCK, UPGRADE_BLOCK, new RemoveDelayHpLoseBlock());
    }

    @Override
    public void use(AbstractPlayer player, AbstractMonster monster) {
        addToBot_gainBlock();
        addToBot(new ChoseCardFromHandCardSelectScreen(
                card -> {
                    addToBot_letSpecificCardExhaust(card);
                    if (!CardUtility.canUseWithoutEnvironment(card))
                        addToBot(new ExhaustSpecificCardAction(card, AbstractDungeon.player.hand));
                })
                .setWindowText(String.format(getEXTENDED_DESCRIPTION()[0], this.magicNumber))
                .setChoiceAmount(this.magicNumber)
                .setRetainFilter(card -> !card.exhaust, card -> !CardModifierManager.hasModifier(card, ExhaustMod.ID))
        );
    }

    public void addToBot_letSpecificCardExhaust(AbstractCard card) {
        AutoDoneInstantAction.addToBotAbstract(() -> {
            card.superFlash();
            CardModifierManager.addModifier(card, new ExhaustMod());
        });
    }

    @Override
    public void upgradeAuto() {
        this.upgradeBaseCost(COST_UPGRADED_NEW);
    }
}

File: superstitioMain/java/superstitio/cards/lupa/SkillCard/cardManipulation/ZenState.kt
package superstitio.cards.lupa.SkillCard.cardManipulation

import basemod.cardmods.ExhaustMod
import basemod.helpers.CardModifierManager
import com.megacrit.cardcrawl.actions.common.ExhaustSpecificCardAction
import com.megacrit.cardcrawl.cards.AbstractCard
import com.megacrit.cardcrawl.characters.AbstractPlayer
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.monsters.AbstractMonster
import superstitio.DataManager
import superstitio.cards.lupa.LupaCard
import superstitio.delayHpLose.RemoveDelayHpLoseBlock
import superstitioapi.actions.AutoDoneInstantAction
import superstitioapi.actions.ChoseCardFromHandCardSelectScreen
import superstitioapi.cards.addExhaustMod
import superstitioapi.utils.CardUtility
import java.util.function.Predicate

class ZenState : LupaCard(ID, CARD_TYPE, COST, CARD_RARITY, CARD_TARGET)
{
    init
    {
        this.setupMagicNumber(MAGIC)
        this.setupBlock(BLOCK, UPGRADE_BLOCK, RemoveDelayHpLoseBlock())
    }

    fun addToBot_letSpecificCardExhaust(card: AbstractCard)
    {
        AutoDoneInstantAction.addToBotAbstract {
            card.superFlash()
            card.addExhaustMod()
        }
    }

    override fun use(player: AbstractPlayer?, monster: AbstractMonster?)
    {
        addToBot_gainBlock()
        ChoseCardFromHandCardSelectScreen { card: AbstractCard ->
            addToBot_letSpecificCardExhaust(card)
            if (!CardUtility.canUseWithoutEnvironment(card)) addToBot(
                ExhaustSpecificCardAction(
                    card,
                    AbstractDungeon.player.hand
                )
            )
        }
            .setWindowText(String.format(cardStrings.getEXTENDED_DESCRIPTION(0), this.magicNumber))
            .setChoiceAmount(this.magicNumber)
            .setRetainFilter(
                Predicate { card: AbstractCard -> !card.exhaust },
                Predicate { card: AbstractCard -> !CardModifierManager.hasModifier(card, ExhaustMod.ID) })
            .addToBot()
    }

    override fun upgradeAuto()
    {
        this.upgradeBaseCost(COST_UPGRADED_NEW)
    }

    companion object
    {
        val ID: String = DataManager.MakeTextID(ZenState::class.java)

        val CARD_TYPE: CardType = CardType.SKILL

        val CARD_RARITY: CardRarity = CardRarity.COMMON

        val CARD_TARGET: CardTarget = CardTarget.SELF

        private const val COST = 0
        private const val COST_UPGRADED_NEW = 0
        private const val BLOCK = 4
        private const val UPGRADE_BLOCK = 2
        private const val MAGIC = 1
    }
}
